<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Software installation | bio.utf8</title>
  <meta name="description" content="" />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Software installation | bio.utf8" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Software installation | bio.utf8" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="bio-data.html"/>
<script src="libs/header-attrs-2.5/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0/anchor-sections.js"></script>

<script src="assets/prism.js"></script>

<link rel="stylesheet" type="text/css" href="assets/bio.css" />

<link rel="stylesheet" type="text/css" href="assets/asciinema-player.css" />

<link rel="stylesheet" type="text/css" href="assets/prism.css" />

<script>

    $(document).ready(function () {
        $('pre').addClass('language-bash');
        Prism.highlightAll();
    });

</script>



</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="index.html">bio: making bioinformatics fun again</a></li>
<li class="divider"></li>
<li><a href="index.html#welcome-to-bio">Welcome to <code>bio</code></a></li>
<li class="chapter" data-level="" data-path="bio-install.html"><a href="bio-install.html"><i class="fa fa-check"></i>Software installation</a></li>
<li class="chapter" data-level="" data-path="bio-data.html"><a href="bio-data.html"><i class="fa fa-check"></i>Fetching data</a></li>
<li class="chapter" data-level="" data-path="bio-fasta.html"><a href="bio-fasta.html"><i class="fa fa-check"></i>Convert to FASTA</a></li>
<li class="chapter" data-level="" data-path="bio-gff.html"><a href="bio-gff.html"><i class="fa fa-check"></i>Convert to GFF</a></li>
<li class="chapter" data-level="" data-path="bio-json.html"><a href="bio-json.html"><i class="fa fa-check"></i>Convert to JSON</a></li>
<li class="chapter" data-level="" data-path="bio-align.html"><a href="bio-align.html"><i class="fa fa-check"></i>Sequence alignments</a></li>
<li class="chapter" data-level="" data-path="bio-taxonomy.html"><a href="bio-taxonomy.html"><i class="fa fa-check"></i>Taxonomy operations</a></li>
<li class="chapter" data-level="" data-path="bio-sra.html"><a href="bio-sra.html"><i class="fa fa-check"></i>Searching SRA</a></li>
<li class="chapter" data-level="" data-path="comm-py.html"><a href="comm-py.html"><i class="fa fa-check"></i>comm.py</a></li>
<li class="chapter" data-level="" data-path="bio-examples.html"><a href="bio-examples.html"><i class="fa fa-check"></i>Usage examples</a></li>
<li class="chapter" data-level="" data-path="bio-tips.html"><a href="bio-tips.html"><i class="fa fa-check"></i>General tips</a></li>
<li class="chapter" data-level="" data-path="bio-related.html"><a href="bio-related.html"><i class="fa fa-check"></i>Related software</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"></a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="bio-install" class="section level1">
<h1>Software installation</h1>
<p><code>bio</code> works on Linux and Mac computers and on Windows when using the Linux Subsystem.</p>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>While the following command should work:</p>
<pre><code>pip install bio --upgrade</code></pre>
<p>We usually recommend installing prerequisites with conda like so:</p>
<pre><code>conda install -c bioconda biopython</code></pre>
<p>then proceed with <code>pip install bio --upgrade</code>.</p>
</div>
<div id="quick-start" class="section level2">
<h2>Quick start</h2>
<p>Run a simple fetch command, set the verbose mode to see what is happening:</p>
<pre><code>bio NC_045512 --fetch -v</code></pre>
<p>now list the known data:</p>
<pre><code>bio --list</code></pre>
<p>try out a conversion:</p>
<pre><code>bio NC_045512 --gff</code></pre>
</div>
<div id="usage" class="section level2">
<h2>Usage</h2>
<p>Type <code>bio</code> followed by one or more accession numbers or data names followed by one or more flags or options.</p>
</div>
<div id="subcommands" class="section level2">
<h2>Subcommands</h2>
<p>Certain flags trigger different behaviors:</p>
<pre><code>bio --align
bio --taxon
bio --sra </code></pre>
</div>
<div id="getting-help" class="section level2">
<h2>Getting help</h2>
<pre><code>bio -h</code></pre>
<p>Subcommands will have separate help pages. For example:</p>
<pre><code>bio --align -h</code></pre>
</div>
<div id="actual-help-pages" class="section level2">
<h2>Actual help pages</h2>
<p>Below we include the help page for each active command</p>
<div id="help-page-for-default-actions" class="section level3">
<h3>1. Help page for default actions</h3>
<pre class="bash"><code>bio -h</code></pre>
<pre><code>usage: bio [-h] [-F] [-G] [-K] [-f] [-u] [-d] [-l] [-P] [-T] [-X] [-E] [-C]
           [-R] [-r &#39;&#39;] [-S &#39;&#39;] [-s &#39;&#39;] [-e &#39;&#39;] [-t &#39;&#39;] [-g &#39;&#39;] [-N &#39;&#39;]
           [-m &#39;&#39;] [-i] [-O] [-v]
           [acc [acc ...]]

bio - making bioinformatics fun again

Command line utilities for manipulating bioinformatics data.

Subcommands with additional help:

    bio --align -h
    bio --taxon -h
    bio --sra -h

positional arguments:
  acc                 data names

optional arguments:
  -h, --help          show this help message and exit
  -F, --fasta         produce FASTA format
  -G, --gff           produce GFF format
  -K, --genbank       show the genbank file if exists
  -f, --fetch         download data as accessions
  -u, --update        updates data in storage
  -d, --delete        delete data in storage
  -l, --list          list data in storage
  -P, --protein       operate on proteins
  -T, --translate     translate DNA to protein
  -X, --transcribe    transrcribe DNA to RNA
  -E, --reverse       reverse sequence
  -C, --complement    complement sequence
  -R, --revcomp       reverse complement sequence
  -r &#39;&#39;, --rename &#39;&#39;  set the name
  -S &#39;&#39;, --seqid &#39;&#39;   set the sequence id
  -s &#39;&#39;, --start &#39;&#39;   start coordinate
  -e &#39;&#39;, --end &#39;&#39;     end coordinate
  -t &#39;&#39;, --type &#39;&#39;    select feature by type
  -g &#39;&#39;, --gene &#39;&#39;    select features associated with gene
  -N &#39;&#39;, --name &#39;&#39;    select features associated with a name
  -m &#39;&#39;, --match &#39;&#39;   select features by rexep match
  -i, --inter         interactive (data from command line)
  -O, --origin        use the origin (source) sequences
  -v, --verbose       verbose mode</code></pre>
</div>
<div id="help-page-for-alignments" class="section level3">
<h3>2. Help page for alignments</h3>
<pre class="bash"><code>bio --align -h</code></pre>
<pre><code>usage: bio [-h] [-s 1] [-e &#39;&#39;] [-o 11] [-x 1] [-L] [-G] [-S] [-p] [-t] [-i]
           [-T] [-R] [-1] [-3] [-v]
           [target] [query]

Performs an alignment between the query and target.

positional arguments:
  target                target sequence to align
  query                 query sequence to align

optional arguments:
  -h, --help            show this help message and exit
  -s 1, --start 1       start coordinate
  -e &#39;&#39;, --end &#39;&#39;       end coordinate
  -o 11, --gap-open 11  gap open penalty
  -x 1, --gap-extend 1  gap extend penalty
  -L, --local           perform local alignment
  -G, --global          perform global alignment (zero end gap penalty)
  -S, --semiglobal      perform a semiglobal alignment
  -p, --protein         use the translated protein sequences from the data
  -t, --translate       translate the DNA into proteins
  -i, --inter           interactive mode, data from command line
  -T, --table           generate an alignment table
  -R, --strict          strict global alignment, apply end gap penalties
  -1, --pep1            shows a translated peptide with one letter code
  -3, --pep3            shows a translated peptide with three letter code
  -v, --verbose         verbose mode, progress messages printed</code></pre>
</div>
<div id="help-page-for-taxonomy" class="section level3">
<h3>3. Help page for taxonomy</h3>
<pre class="bash"><code>bio --taxon -h</code></pre>
<pre><code>usage: bio [-h] [-T 0] [-l] [-f] [-i   ] [-S ,] [-L] [-b] [-d] [-p] [-v]
           [words [words ...]]

positional arguments:
  words                 taxids or search queries

optional arguments:
  -h, --help            show this help message and exit
  -T 0, --limit 0       limit the number of entries
  -l, --list            lists database content
  -f, --flat            flattened output
  -i    , --indent      the indentation string
  -S , , --sep ,        separator string
  -L, --lineage         show the lineage for a taxon term
  -b, --build           build a database from a taxdump
  -d, --download        download newest taxdump from NCBI
  -p, --preload         loads entire database in memory
  -v, --verbose         verbose mode, prints more messages</code></pre>
</div>
<div id="help-page-for-sra-search" class="section level3">
<h3>4. Help page for SRA search</h3>
<pre class="bash"><code>bio --sra -h</code></pre>
<pre><code>usage: bio [-h] [-p] [-l &#39;&#39;] [-s] [-t] [-i] [-v] [acc [acc ...]]

positional arguments:
  acc                accessions

optional arguments:
  -h, --help         show this help message and exit
  -p, --project      project run information
  -l &#39;&#39;, --limit &#39;&#39;  limit the number of results
  -s, --sample       sample information
  -t, --table        tabular output
  -i, --inter        interactive (data from command line)
  -v, --verbose      verbose mode, prints more messages</code></pre>

</div>
</div>
</div>

<script src="assets/asciinema-player.js"></script>

            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="bio-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": null,
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "none"
}
});
});
</script>

</body>

</html>
