<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Fetching data | bio.utf8</title>
  <meta name="description" content="" />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Fetching data | bio.utf8" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Fetching data | bio.utf8" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="bio-install.html"/>
<link rel="next" href="bio-fasta.html"/>
<script src="libs/header-attrs-2.5/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0/anchor-sections.js"></script>

<script src="assets/prism.js"></script>

<link rel="stylesheet" type="text/css" href="assets/bio.css" />

<link rel="stylesheet" type="text/css" href="assets/asciinema-player.css" />

<link rel="stylesheet" type="text/css" href="assets/prism.css" />

<script>

    $(document).ready(function () {
        $('pre').addClass('language-bash');
        Prism.highlightAll();
    });

</script>



</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="index.html">bio: making bioinformatics fun again</a></li>
<li class="divider"></li>
<li><a href="index.html#welcome-to-bio">Welcome to <code>bio</code></a></li>
<li class="chapter" data-level="" data-path="bio-install.html"><a href="bio-install.html"><i class="fa fa-check"></i>Software installation</a></li>
<li class="chapter" data-level="" data-path="bio-data.html"><a href="bio-data.html"><i class="fa fa-check"></i>Fetching data</a></li>
<li class="chapter" data-level="" data-path="bio-fasta.html"><a href="bio-fasta.html"><i class="fa fa-check"></i>Convert to FASTA</a></li>
<li class="chapter" data-level="" data-path="bio-gff.html"><a href="bio-gff.html"><i class="fa fa-check"></i>Convert to GFF</a></li>
<li class="chapter" data-level="" data-path="bio-json.html"><a href="bio-json.html"><i class="fa fa-check"></i>Convert to JSON</a></li>
<li class="chapter" data-level="" data-path="bio-align.html"><a href="bio-align.html"><i class="fa fa-check"></i>Sequence alignments</a></li>
<li class="chapter" data-level="" data-path="bio-taxonomy.html"><a href="bio-taxonomy.html"><i class="fa fa-check"></i>Taxonomy operations</a></li>
<li class="chapter" data-level="" data-path="bio-sra.html"><a href="bio-sra.html"><i class="fa fa-check"></i>Searching SRA</a></li>
<li class="chapter" data-level="" data-path="bio-examples.html"><a href="bio-examples.html"><i class="fa fa-check"></i>Usage examples</a></li>
<li class="chapter" data-level="" data-path="bio-tips.html"><a href="bio-tips.html"><i class="fa fa-check"></i>General tips</a></li>
<li class="chapter" data-level="" data-path="bio-related.html"><a href="bio-related.html"><i class="fa fa-check"></i>Related software</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"></a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="bio-data" class="section level1">
<h1>Fetching data</h1>
<p>The <code>bio</code> package solves the ongoing struggle of how to maintain sanity among diverse datasets. When you obtain data with <code>bio</code> it becomes universally available to all tools in the package.</p>
<div id="getting-data-from-ncbi-fetch" class="section level2">
<h2>Getting data from NCBI (–fetch)</h2>
<p>The <code>--fetch</code> command downloads data identified via accession numbers from NCBI then stores
this data in a storage directory (<code>~/.bio</code>). All subsequent commands in the <code>bio</code> package can seamlessly access the stored data from any location and would not need to connect to the internet to use it.</p>
<pre><code># Run fetch in verbose mode.
bio NC_045512 --fetch -v</code></pre>
<p>Running the fetch command the next time for the same accession number will not connect to the internet again, it will use the existing data instead. Use the <code>--fetch --update</code> (see later) to force a re-downloading of data from NCBI.</p>
<p>Most commands can operate on multiple accession numbers at a time.</p>
<pre><code>bio NC_045512 MN996532 --fetch</code></pre>
<p>There will be commands like <code>--rename</code> where it makes no sense to apply the operation on multiple data at the same time. In those cases only the first accession number is acted upon.</p>
<p>The internal, gzip compressed, JSON based representation used by <code>bio</code> is simple, efficient and highly performant. A 330MB GenBank file for chromosome 1 of the human genome obtained from NCBI will in our data representation take just 67MB to store. More importantly our data representation can read and converted to <code>fasta</code> format in 6 seconds:</p>
<pre><code>time bio NC_000001 --fasta | wc -c
253105766

real    0m6.189s</code></pre>
</div>
<div id="rename-rename" class="section level2">
<h2>Rename (–rename)</h2>
<p>Accession numbers are tedious to handle. Almost always we rename data to be meaningful.</p>
<pre><code>bio NC_045512 --fetch --rename ncov</code></pre>
<p>the command above will store the data under the name <code>ncov</code>. Within the data the sequence will still be labeled as <code>NC_045512</code>. You may change both the name and sequence id:</p>
<pre><code>bio NC_045512 --fetch --rename ncov --seqid ncov
bio NC_045512 --fetch --rename ratg13 --seqid ratg13</code></pre>
<p>Now, not only is your data called <code>ncov</code> but the sequence id inside the data is also set to <code>ncov</code>.</p>
</div>
<div id="listing-the-storage-list" class="section level2">
<h2>Listing the storage (–list)</h2>
<p>Each time data is fetched is will be stored locally. The <code>bio</code> package will look in this storage no matter what directory you run it from. To list the content of the storage write:</p>
<pre><code>bio --list</code></pre>
<p>it prints:</p>
<pre><code>22K   ncov     Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome
19K   ratg13   Bat coronavirus RaTG13, complete genome</code></pre>
</div>
<div id="delete-data-delete" class="section level2">
<h2>Delete data (–delete)</h2>
<p>To drop data from storage use:</p>
<pre><code>bio ncov --delete</code></pre>
<p>This command will only drop the JSON representation not the downloaded GenBank if exists.
If you want to update the original data use the <code>--update</code> parameter.</p>
</div>
<div id="update-data-update" class="section level2">
<h2>Update data (–update)</h2>
<p>To force fetch to download data that already seems to be present do:</p>
<pre><code>bio NC_045512 --fetch --update</code></pre>
<p>Note that you can’t update a renamed sequence. At that point the original accession number is lost. You can however fetch, update and rename all in one go like so:</p>
<pre><code>bio NC_045512 --fetch --update --rename ncov --seqid ncov</code></pre>
<p>There is a builtin order of operations, does not matter what order you list commands. For example <code>--delete</code> would take place first before the <code>--fetch</code> and so on.</p>
</div>
<div id="view-data" class="section level2">
<h2>View data</h2>
<p>The default action is to view the stored data. Locally the data is stored in a JSON format that makes processing it much faster than the original GenBank yet has no loss of information:</p>
<pre><code>bio ncov | head </code></pre>
</div>
<div id="local-data" class="section level2">
<h2>Local data</h2>
<p><code>bio</code> can also read and process data from local files. If you have a genbank or fasta file you can use that as input. Here we turn a genbank file into gff:</p>
<pre><code>bio  mydata.gb --gff </code></pre>
</div>
<div id="data-from-command-line" class="section level2">
<h2>Data from command line</h2>
<p>In addition, there is a so called <em>interactive</em> input of data (<code>-i</code>) where the data can be listed at the command line:</p>
<pre><code>bio ATGAATATATAC -i --translate</code></pre>
<p>The above command will operate on the sequence as if it were stored in a FASTA file, the above prints:</p>
<pre><code>&gt;S1 translated DNA
MNIY</code></pre>
</div>
<div id="what-is-the-point-of-getting-data-from-command-line" class="section level2">
<h2>What is the point of getting data from command line?</h2>
<p>It is an explicit way to explore and demonstrate information. For example, suppose you wanted to see how the same DNA sequence would be
translated to different amino acides when using the the first or second reading frame:</p>
<pre class="bash"><code>bio ATGAATATATACT -i --translate --start 1</code></pre>
<pre><code>&gt;S1 translated
MNIY</code></pre>
<p>versus</p>
<pre class="bash"><code>bio ATGAATATATACT -i --translate --start 2</code></pre>
<pre><code>&gt;S1 [2:13], translated
*IYT</code></pre>
<p>or you can explore alignments:</p>
<pre><code>bio THISLINE ISALIGNED --align -i </code></pre>
<p>to see:</p>
<pre><code>S1           THISLI--NE-
           1   ||.:  ||  11
S2           --ISALIGNED</code></pre>

</div>
</div>

<script src="assets/asciinema-player.js"></script>

            </section>

          </div>
        </div>
      </div>
<a href="bio-install.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="bio-fasta.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": null,
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "none"
}
});
});
</script>

</body>

</html>
