<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Data access | bio.utf8</title>
  <meta name="description" content="" />
  <meta name="generator" content="bookdown 0.18 and GitBook 2.6.7" />

  <meta property="og:title" content="Data access | bio.utf8" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Data access | bio.utf8" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="install.html"/>
<link rel="next" href="convert.html"/>
<script src="libs/header-attrs-2.1/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />










<script src="assets/prism.js"></script>

<link rel="stylesheet" type="text/css" href="assets/bio.css" />

<link rel="stylesheet" type="text/css" href="assets/asciinema-player.css" />

<link rel="stylesheet" type="text/css" href="assets/prism.css" />

<script>

    $(document).ready(function () {
        $('pre').addClass('language-bash');
        Prism.highlightAll();
    });

</script>



</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="index.html">bio: making bioinformatics fun again</a></li>
<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="install.html"><a href="install.html"><i class="fa fa-check"></i>Installation</a></li>
<li class="chapter" data-level="" data-path="data.html"><a href="data.html"><i class="fa fa-check"></i>Data access</a></li>
<li class="chapter" data-level="" data-path="convert.html"><a href="convert.html"><i class="fa fa-check"></i>Format conversions</a></li>
<li class="chapter" data-level="" data-path="align.html"><a href="align.html"><i class="fa fa-check"></i>Sequence alignments</a></li>
<li class="chapter" data-level="" data-path="align-the-extracted-protein-.html"><a href="align-the-extracted-protein-.html"><i class="fa fa-check"></i>Align the extracted protein.</a></li>
<li class="chapter" data-level="" data-path="align-the-extracted-protein-1.html"><a href="align-the-extracted-protein-1.html"><i class="fa fa-check"></i>Align the extracted protein.</a></li>
<li class="chapter" data-level="" data-path="align-the-translated-regions-.html"><a href="align-the-translated-regions-.html"><i class="fa fa-check"></i>Align the translated regions.</a></li>
<li class="chapter" data-level="" data-path="test-alignments.html"><a href="test-alignments.html"><i class="fa fa-check"></i>Test alignments</a></li>
<li class="chapter" data-level="" data-path="default-alignment-is-local-.html"><a href="default-alignment-is-local-.html"><i class="fa fa-check"></i>Default alignment is local.</a></li>
<li class="chapter" data-level="" data-path="global-alignment-.html"><a href="global-alignment-.html"><i class="fa fa-check"></i>Global alignment.</a></li>
<li class="chapter" data-level="" data-path="semiglobal-alignment-.html"><a href="semiglobal-alignment-.html"><i class="fa fa-check"></i>Semiglobal alignment.</a></li>
<li class="chapter" data-level="" data-path="taxon.html"><a href="taxon.html"><i class="fa fa-check"></i>NCBI Taxonomy</a></li>
<li class="chapter" data-level="" data-path="tips.html"><a href="tips.html"><i class="fa fa-check"></i>Usage tips</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"></a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="data" class="section level1">
<h1>Data access</h1>
<p>The <code>bio</code> package solves the ongoing struggle of how to maintain sanity among diverse datasets.</p>
<p>When you obtain data with <code>bio</code> it becomes universally available to all tools in the package.</p>
<div id="local-data" class="section level3">
<h3>Local data</h3>
<p>There is an automated data storage in <code>bio</code> that makes using data from NCBI quite convenient. But before we even go ther let’s make it clear that <code>bio</code> can read and process data from local files just fine. If you have a genbank or fasta file you can use that as input. Here we turn a genbank file into gff:</p>
<pre><code>bio  mydata.gb --gff </code></pre>
<p>Moreover there is a so called command line input of data (<code>-i</code>) where the data may be listed at the command line:</p>
<pre><code>bio ATGAATATATAC -i --translate</code></pre>
<p>The above command will operate on the sequence as if it were stored in a FASTA file to produce:</p>
<pre><code>&gt;S1 translated DNA
MNIY</code></pre>
</div>
<div id="getting-data-from-ncbi-fetch" class="section level3">
<h3>Getting data from NCBI (–fetch)</h3>
<p>The <code>--fetch</code> command downloads data identified via accession numbers from NCBI then stores
this data in a storage directory (<code>~/.bio</code>). All subsequent commands in the <code>bio</code> package can seamlessly access the stored data from any location and would not need to connect to the internet to use it.</p>
<pre><code># Run fetch in verbose mode.
bio NC_045512 --fetch -v</code></pre>
<p>Running the fetch command the next time for the accession number will not connect to the internet again, it will exit instead. Use the <code>--fetch --update</code> (see later) to force a re-downloading of data from NCBI.</p>
<p>Most commands can operate on multiple accession numbers at a time.</p>
<pre><code>bio NC_045512 MN996532 --fetch</code></pre>
<p>There will be commands like <code>--rename</code> where it makes no sense to apply the operation on multiple data at the same time. In those cases only the first accession number is acted upon.</p>
</div>
<div id="data-performance" class="section level2">
<h2>Data performance</h2>
<p>All fetched data is stored in a compressed JSON format. A few observed benchmarks that will heavily depend on NCBI performance:</p>
<ul>
<li><code>NC_045512</code>, SARS-COV-2 virus: 2 seconds, 1,200 lines, 77KB downloaded, 22KB when stored.</li>
<li><code>NC_002695</code>, Escherichia coli bacteria: 1 minute, 170,000 lines, 11MB downloaded, 2.7MB when stored.</li>
<li><code>NT_033779</code>, Fruit fly chromosome 2L: 2 minutes, 660,000 lines, 45MB downloaded, 12MB when stored.</li>
<li><code>NC_000001</code>, Human genome chromosome 1: 12 minutes, 4,920,000 lines, 300MB downloaded, 64MB when stored.</li>
</ul>
<p>Note that the internal representation is both efficient and speedy. Even the largest of files takes just a few seconds to be converted to say <code>fasta</code> format.:</p>
<pre><code>time bio NC_000001 --fasta | wc -l
4149276

real    0m6.189s
user    0m3.844s
sys     0m2.359s</code></pre>
<div id="rename-rename" class="section level3">
<h3>Rename (–rename)</h3>
<p>Accession numbers are tedious to handle. Almost always we rename data to be meaningful.</p>
<pre><code>bio NC_045512 --fetch --rename ncov</code></pre>
<p>the command above will store the data under the name <code>ncov</code>. Within the data the sequence will still be labeled as <code>NC_045512</code>. You may change both the name and sequence id:</p>
<pre><code>bio NC_045512 --fetch --rename ncov --seqid ncov
bio NC_045512 --fetch --rename ratg13 --seqid ratg13</code></pre>
<p>Now, not only is your data called <code>ncov</code> but the sequence id inside the data is also set to <code>ncov</code>.</p>
</div>
<div id="listing-the-storage-list" class="section level3">
<h3>Listing the storage (–list)</h3>
<p>Each time data is fetched is will be stored locally. The <code>bio</code> package will look in this storage no matter what directory you run it from. To list the content of the storage write:</p>
<pre><code>bio --list</code></pre>
<p>it prints:</p>
<pre><code>22K   ncov     Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome
19K   ratg13   Bat coronavirus RaTG13, complete genome</code></pre>
</div>
<div id="delete-data-delete" class="section level3">
<h3>Delete data (–delete)</h3>
<p>To drop data from storage use:</p>
<pre><code>bio ncov --delete</code></pre>
<p>This command will only drop the JSON representation not the downloaded GenBank if exists.
If you want to update the original data use the <code>--update</code> parameter.</p>
</div>
<div id="update-data-update" class="section level3">
<h3>Update data (–update)</h3>
<p>To force fetch to download data that already seems to be present do:</p>
<pre><code>bio NC_045512 --fetch --update</code></pre>
<p>Note that you can’t update a renamed sequence. At that point the original accession number is lost. You can however fetch, update and rename all in one go like so:</p>
<pre><code>bio NC_045512 --fetch --update --rename ncov --seqid ncov</code></pre>
<p>There is a builtin order of operations, does not matter what order you list commands. For example <code>--delete</code> would take place first before the <code>--fetch</code> and so on.</p>
</div>
<div id="view-data" class="section level3">
<h3>View data</h3>
<p>The default action is to view the stored data. Locally the data is stored in a JSON format that makes processing it much faster than the original GenBank yet has no loss of information:</p>
<pre><code>bio ncov | head </code></pre>
<p>You can subselect and view sections of the data, for example the <code>S</code> gene:</p>
<pre><code>bio ncov:S | head</code></pre>
<p>Many other combinations are valid:</p>
<pre><code>bio ncov --type CDS </code></pre>
<p>When viewing the data directly you are shown the JSON based data that <code>bio</code> uses internally.</p>
</div>
</div>
<div id="example-output" class="section level2">
<h2>Example output</h2>
<pre class="bash"><code>bio ncov | head</code></pre>
<pre><code>[
    {
        &quot;id&quot;: &quot;ncov&quot;,
        &quot;definition&quot;: &quot;Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome&quot;,
        &quot;dblink&quot;: [
            &quot;BioProject:PRJNA485481&quot;
        ],
        &quot;locus&quot;: &quot;NC_045512&quot;,
        &quot;feature_count&quot;: 57,
        &quot;origin_len&quot;: 29903,</code></pre>

</div>
</div>

<script src="assets/asciinema-player.js"></script>

            </section>

          </div>
        </div>
      </div>
<a href="install.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="convert.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": null,
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "none"
}
});
});
</script>

</body>

</html>
