<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Usage examples | bio.utf8</title>
  <meta name="description" content="" />
  <meta name="generator" content="bookdown 0.18 and GitBook 2.6.7" />

  <meta property="og:title" content="Usage examples | bio.utf8" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Usage examples | bio.utf8" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="taxon.html"/>
<link rel="next" href="tips.html"/>
<script src="libs/header-attrs-2.1/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />










<script src="assets/prism.js"></script>

<link rel="stylesheet" type="text/css" href="assets/bio.css" />

<link rel="stylesheet" type="text/css" href="assets/asciinema-player.css" />

<link rel="stylesheet" type="text/css" href="assets/prism.css" />

<script>

    $(document).ready(function () {
        $('pre').addClass('language-bash');
        Prism.highlightAll();
    });

</script>



</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="index.html">bio: making bioinformatics fun again</a></li>
<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome to bioinformatics</a></li>
<li class="chapter" data-level="" data-path="install.html"><a href="install.html"><i class="fa fa-check"></i>Software installation</a></li>
<li class="chapter" data-level="" data-path="data.html"><a href="data.html"><i class="fa fa-check"></i>Data management</a></li>
<li class="chapter" data-level="" data-path="convert.html"><a href="convert.html"><i class="fa fa-check"></i>Format conversions</a></li>
<li class="chapter" data-level="" data-path="align.html"><a href="align.html"><i class="fa fa-check"></i>Sequence alignments</a></li>
<li class="chapter" data-level="" data-path="taxon.html"><a href="taxon.html"><i class="fa fa-check"></i>Taxonomy operations</a></li>
<li class="chapter" data-level="" data-path="examples.html"><a href="examples.html"><i class="fa fa-check"></i>Usage examples</a></li>
<li class="chapter" data-level="" data-path="tips.html"><a href="tips.html"><i class="fa fa-check"></i>General tips</a></li>
<li class="chapter" data-level="" data-path="related.html"><a href="related.html"><i class="fa fa-check"></i>Related software</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"></a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="examples" class="section level1">
<h1>Usage examples</h1>
<p>The following script is the test file that we use to verify the various functionalities of <code>bio</code></p>
<pre class="bash"><code>#
# This script is used to generate Python tests.
#
# The output generated by each test can be seen at:
#
# https://github.com/ialbert/bio/tree/master/test/data
#

# Stop on errors.
set -uex

# Delete the ncov data if exists.
bio ncov --delete

# Fetch the accession, rename the data and change the sequence id.
bio NC_045512 --fetch --rename ncov --seqid ncov

# You can also fetch and convert at the same time.
bio NC_045512 --fetch --fasta --end 100 &gt; fetch-convert.fa

# Shows the internal JSON format of the data.
bio ncov &gt; ncov.json

# Convert to GenBank.
bio ncov --genbank &gt; ncov.gb

# Convert to FASTA.
bio ncov --fasta &gt; ncov.fa

# Convert to GFF.
bio ncov --gff &gt; ncov.gff

# Filter the internal JSON by regular expression match and type.
bio ncov --match ORF1ab --type gene &gt; match.json

# Convert to GFF features associated with a gene.
bio ncov --gff --gene S &gt; gene1.gff

# Convert to GFF features that overlap with start to end.
bio ncov --gff  --start 10000 --end 20000 &gt; overlap.gff

# Numbers may have commas.
bio ncov --gff  --start 10,000 --end 20,000 &gt; overlap.gff

# Numbers may be expressed with prefixes.
bio ncov --gff  --start 10kb --end 20kb &gt; overlap.gff

# Convert to GFF by type.
bio ncov --gff  --type CDS &gt; type.gff

# Convert to GFF by multiple types.
bio ncov --gff  --type gene,CDS,mRNA &gt; multiple-types.gff

# Slice a FASTA to a region and change the sequence id.
bio ncov --fasta --seqid foo --start 10 --end 20 &gt; fasta-start.fa

# Convert to FASTA features with a certain type.
bio ncov --fasta --type CDS &gt; CDS.fa

# Convert to FASTA a sub region of type filtered data.
bio ncov --fasta --type gene --end 10 &gt; gene-start.fa

# Translate the DNA for features that have the type CDS.
bio ncov --translate --type CDS &gt; translate.fa

# Extract already(!) translated proteins from the data.
# The translation attribute must be filled in GenBank.
bio ncov --protein --start -10 &gt; protein-end.fa

# There are shortcuts to accesing coding sequences for genes.
# The following two commands are equivalent.
bio ncov --fasta --type CDS --gene S --end 10 &gt; shortcut.fa
bio ncov:S --fasta --end 10 &gt; shortcut.fa

# Extract the already traslated protein from the data.
bio ncov:S --fasta --protein --seqid foo &gt; s_prot_foo.fa

# Interactive mode. Data obtained from the command line paramter
bio ATGGGC -i --fasta &gt; inter.fa

# Translate in interactive mode.
bio ATGGGC -i --translate &gt;  inter-trans.fa

# Translate on the reverse complement.
bio ATGGGC -i --revcomp --translate &gt; inter-revcomp1.fa

# You can separately reverse and complement
bio ATGGGC -i --reverse --complement --translate &gt;  inter-revcomp2.fa

# Get the RaTG13 data.
bio MN996532 --fetch --rename ratg13 --seqid ratg13

# Align the first 200 bp across both genomes.
bio ncov ratg13 --end 210 --align &gt; align-dna.txt

# Align the DNA for the coding sequences of the S protein.
bio ncov:S ratg13:S --end 210 --align &gt; align-dna-s.txt

# Align the translated DNA for the coding sequences of the S protein.
# Slice appled to DNA before translation.
bio ncov:S ratg13:S --end 210 --translate --align &gt; align-translated-s.txt

# Align the already translated proteins.
# Slice applied to the protein sequence.
bio ncov:S ratg13:S --protein --end 70 --align &gt; align-protein-s.txt

# Local alignment in interactive mode.
bio THISLINE ISALIGNED  -i --align --local &gt; align-local.txt

# Global alignment in interactive mode.
bio THISLINE ISALIGNED -i --align --global &gt; align-global.txt

# Semiglobal alignment in interactive mode.
bio THISLINE ISALIGNED -i --align --semiglobal &gt; align-semiglobal.txt

# Check taxonomy defaults
bio 9606 --taxon &gt; taxon_9606.txt

# Generate lineage.
bio 9606 --lineage --taxon &gt; taxon_9606_lineage.txt

# The lineage may be flattened to a single line.
bio 9606 --lineage --flat --taxon &gt; taxon_9606_flat_lineage.txt

# Taxonomy information from data.
bio ncov ratg13 --taxon &gt; taxon_ncov_ratg13.txt
</code></pre>

</div>

<script src="assets/asciinema-player.js"></script>

            </section>

          </div>
        </div>
      </div>
<a href="taxon.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="tips.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": null,
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "none"
}
});
});
</script>

</body>

</html>
