<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Welcome to bio | bio.utf8</title>
  <meta name="description" content="" />
  <meta name="generator" content="bookdown 0.18 and GitBook 2.6.7" />

  <meta property="og:title" content="Welcome to bio | bio.utf8" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Welcome to bio | bio.utf8" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  

<link rel="next" href="bio-install.html"/>
<script src="libs/header-attrs-2.5/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0/anchor-sections.js"></script>

<script src="assets/prism.js"></script>

<link rel="stylesheet" type="text/css" href="assets/bio.css" />

<link rel="stylesheet" type="text/css" href="assets/asciinema-player.css" />

<link rel="stylesheet" type="text/css" href="assets/prism.css" />

<script>

    $(document).ready(function () {
        $('pre').addClass('language-bash');
        Prism.highlightAll();
    });

</script>



</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="index.html">bio: making bioinformatics fun again</a></li>
<li class="divider"></li>
<li><a href="index.html#welcome-to-bio">Welcome to <code>bio</code></a></li>
<li class="chapter" data-level="" data-path="bio-install.html"><a href="bio-install.html"><i class="fa fa-check"></i>Software installation</a></li>
<li class="chapter" data-level="" data-path="bio-data.html"><a href="bio-data.html"><i class="fa fa-check"></i>Fetching data</a></li>
<li class="chapter" data-level="" data-path="bio-fasta.html"><a href="bio-fasta.html"><i class="fa fa-check"></i>Convert to FASTA</a></li>
<li class="chapter" data-level="" data-path="bio-gff.html"><a href="bio-gff.html"><i class="fa fa-check"></i>Convert to GFF</a></li>
<li class="chapter" data-level="" data-path="bio-json.html"><a href="bio-json.html"><i class="fa fa-check"></i>Convert to JSON</a></li>
<li class="chapter" data-level="" data-path="bio-align.html"><a href="bio-align.html"><i class="fa fa-check"></i>Sequence alignments</a></li>
<li class="chapter" data-level="" data-path="bio-taxonomy.html"><a href="bio-taxonomy.html"><i class="fa fa-check"></i>Taxonomy operations</a></li>
<li class="chapter" data-level="" data-path="bio-sra.html"><a href="bio-sra.html"><i class="fa fa-check"></i>Searching SRA</a></li>
<li class="chapter" data-level="" data-path="bio-examples.html"><a href="bio-examples.html"><i class="fa fa-check"></i>Usage examples</a></li>
<li class="chapter" data-level="" data-path="bio-tips.html"><a href="bio-tips.html"><i class="fa fa-check"></i>General tips</a></li>
<li class="chapter" data-level="" data-path="bio-related.html"><a href="bio-related.html"><i class="fa fa-check"></i>Related software</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"></a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<!--bookdown:title:end-->
<!--bookdown:title:start-->
<div id="welcome-to-bio" class="section level1">
<h1>Welcome to <code>bio</code></h1>
<blockquote>
<p>The software is currently under development. It is operational but not fully vetted.</p>
</blockquote>
<p><code>bio</code> - command-line utilities to make bioinformatics explorations more enjoyable.</p>
<p>Built on top of <a href="https://biopython.org/">BioPython</a>, <a href="https://github.com/jeffdaily/parasail">Parasail</a> and other existing packages; <code>bio</code> streamlines tedious bioinformatics tasks such as:</p>
<ul>
<li>downloading and storing data</li>
<li>converting between formats</li>
<li>extracting a subset of information</li>
<li>visualizing alignments</li>
<li>exploring taxonomies</li>
<li>… and more …</li>
</ul>
<p>Having access to all the utility described above makes the <code>bio</code> package well suited for exploratory analysis of genomes. The software was written to teach bioinformatics and is the companion software to the <a href="https://www.biostarhandbook.com/">Biostar Handbook</a></p>
<div id="quick-links" class="section level2">
<h2>Quick links</h2>
<ul>
<li>Source code: <a href="https://github.com/ialbert/bio" class="uri">https://github.com/ialbert/bio</a></li>
<li>Use cases: <a href="https://github.com/ialbert/bio/blob/master/test/bio_examples.sh">bio_examples.sh</a></li>
</ul>
</div>
<div id="why-do-we-need-this-software" class="section level2">
<h2>Why do we need this software?</h2>
<p>If you’ve ever done bioinformatics you know how even seemingly straightforward tasks require multiple steps, arcane incantations, reading documentation and numerous other preparations that slow down your progress.</p>
<p>Time and again I found myself not pursuing an idea because getting to the fun part was too tedious. The <code>bio</code> package is meant to solve that tedium.</p>
</div>
<div id="diving-right-in" class="section level2">
<h2>Diving right in</h2>
<p>Here is how to align the first 1000 basepairs of SARS-COV-2 versus to the same region of a bat coronavirus:</p>
<pre><code># Get the data.
bio NC_045512 MN996532 --fetch 

# Align the sequences.
bio NC_045512 MN996532 --align --end 1000</code></pre>
<p>that’s it. <code>bio</code> will take care of everything for you and prints:</p>
<pre><code>### 1: NC_045512.2 vs MN996532.2 ###

Length: 1000 (semiglobal)
Query:  1000 [1, 1000]
Target: 1000 [1, 1000]
Score:  4721
Ident:  969/1000 (96.9%)
Simil:  969/1000 (96.9%)
Gaps:   0/1000 (0.0%)
Matrix: nuc44(-11, -1)

NC_045512.2  ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTGGCTGTCACTC
           1 ||||||||||||||||||.|||||||||||||||||.||||.||||||||||||||||||||||||||||||||||||||||||||||||.||||||||| 100
MN996532.2   ATTAAAGGTTTATACCTTTCCAGGTAACAAACCAACGAACTCTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTGACTGTCACTC

NC_045512.2  GGCTGCATGCTTAGTGCACTCACGCAGTATAATTAATAACTAATTACTGTCGTTGACAGGACACGAGTAACTCGTCTATCTTCTGCAGGCTGCTTACGGT
         101 |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||.|||||||||||||||.|||||||||| 200
MN996532.2   GGCTGCATGCTTAGTGCACTCACGCAGTATAATTAATAACTAATTACTGTCGTTGACAGGACACGAGTAACTCATCTATCTTCTGCAGGTTGCTTACGGT

...    </code></pre>
</div>
<div id="a-more-realistic-example" class="section level2">
<h2>A more realistic example</h2>
<p>Suppose you wanted to identify the differences between the <code>S</code> protein of the bat coronavirus deposited as <code>MN996532</code> and the <code>S</code> protein of the ancestral SARS-COV-2 virus designated by the NCBI via accession number <code>NC_045512</code>.</p>
<p>If you are a trained bioinformatician, think about all the steps you would need to perform to accomplish this task, then think about the effort it would take you to teach someone else how to do the same.</p>
<p>With the <code>bio</code> package the process takes simple, concise steps.</p>
</div>
<div id="download-and-rename" class="section level2">
<h2>Download and rename</h2>
<p>First we download and rename the data keep our sanity:</p>
<pre><code>bio NC_045512 --fetch --rename ncov
bio MN996532  --fetch --rename ratg13</code></pre>
<p>From now on <code>bio</code> can operate on <code>NC_045512</code> using the name <code>ncov</code> and on <code>MN996532</code> using the name <code>ratg13</code> no matter where you are on your computer!</p>
</div>
<div id="convert-to-different-formats" class="section level2">
<h2>Convert to different formats</h2>
<p><code>bio</code> stores data in an internal storage system that it can find from any location. There is no clutter of files or paths to remember. For example, in any directory you now can type:</p>
<pre><code>bio ncov --fasta --end 100</code></pre>
<p>and it will show you the first 100 bases of the genome</p>
<pre><code>&gt;ncov Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome
ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGTAGATCT
GTTCTCTAAACGAACTTTAAAATCTGTGTGGCTGTCACTC</code></pre>
<p>You could also convert the data stored under <code>ncov</code> name to other formats. Let’s convert features with type <code>CDS</code> to <code>GFF</code>:</p>
<pre class="bash"><code>bio ncov --gff --type CDS  | head -5</code></pre>
<pre><code>##gff-version 3
ncov    .   CDS 266 13468   .   +   .   ID=CDS-1;Parent=YP_009724389.1;Name=YP_009724389.1;gene=ORF1ab;locus_tag=GU280_gp01;ribosomal_slippage=;note=pp1ab; translated by -1 ribosomal frameshift;codon_start=1;product=ORF1ab polyprotein;protein_id=YP_009724389.1;db_xref=GeneID:43740578;gene_id=GU280_gp01
ncov    .   CDS 13468   21555   .   +   .   ID=CDS-2;Parent=YP_009724389.1;Name=YP_009724389.1;gene=ORF1ab;locus_tag=GU280_gp01;ribosomal_slippage=;note=pp1ab; translated by -1 ribosomal frameshift;codon_start=1;product=ORF1ab polyprotein;protein_id=YP_009724389.1;db_xref=GeneID:43740578;gene_id=GU280_gp01
ncov    .   CDS 266 13483   .   +   .   ID=CDS-3;Parent=YP_009725295.1;Name=YP_009725295.1;gene=ORF1ab;locus_tag=GU280_gp01;note=pp1a;codon_start=1;product=ORF1a polyprotein;protein_id=YP_009725295.1;db_xref=GeneID:43740578;gene_id=GU280_gp01
ncov    .   CDS 21563   25384   .   +   .   ID=CDS-4;Parent=YP_009724390.1;Name=YP_009724390.1;gene=S;locus_tag=GU280_gp02;gene_synonym=spike glycoprotein;note=structural protein; spike protein;codon_start=1;product=surface glycoprotein;protein_id=YP_009724390.1;db_xref=GeneID:43740568;gene_id=GU280_gp02</code></pre>
</div>
<div id="align-nucleotides-or-peptides" class="section level2">
<h2>Align nucleotides or peptides</h2>
<p>Now, back to our problem of aligning proteins. Let’s align the first 90 basepairs of DNA sequences for the <code>S</code> protein for each organism, <code>bio</code> even gives you a shortcut, instead of typing <code>--gene S --type CDS</code> you can write it as <code>ncov:S</code> :</p>
<pre class="bash"><code>bio ncov:S ratg13:S --end 90 --align</code></pre>
<pre><code>
### 1: YP_009724390 vs QHR63300.2 ###

Length: 90 (semiglobal) 
Query:  90 [1, 90]
Target: 90 [1, 90]
Score:  387
Ident:  83/90 (92.2%)
Simil:  83/90 (92.2%)
Gaps:   0/90 (0.0%)
Matrix: nuc44(-11, -1) 

YP_009724390 ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTGTTAATCTTACAACCAGAACTCAATTACCCCCTGCATACACTAAT
           1 ||||||||||||||||||||||||||||||||.||||||||||||||||||||.|||||.||||||||.|||||.|||||||||||.||. 90
QHR63300.2   ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTTTCTAGTCAGTGTGTTAATCTAACAACTAGAACTCAGTTACCTCCTGCATACACCAAC</code></pre>
<p>If instead we wanted to align the 90bp DNA sequences for <code>S</code> protein after their translation into proteins we could do it like so:</p>
<pre class="bash"><code>bio ncov:S ratg13:S --translate --end 90 --align</code></pre>
<pre><code>
### 1: YP_009724390 vs QHR63300.2 ###

Length: 30 (semiglobal) 
Query:  30 [1, 30]
Target: 30 [1, 30]
Score:  153
Ident:  30/30 (100.0%)
Simil:  30/30 (100.0%)
Gaps:   0/30 (0.0%)
Matrix: blosum62(-11, -1) 

YP_009724390 MFVFLVLLPLVSSQCVNLTTRTQLPPAYTN
           1 |||||||||||||||||||||||||||||| 30
QHR63300.2   MFVFLVLLPLVSSQCVNLTTRTQLPPAYTN</code></pre>
<p>We can note right away that all differences in the first 90bp of DNA are synonymous substitutions, the protein translations are the same.</p>
</div>
<div id="look-up-the-taxonomy" class="section level2">
<h2>Look up the taxonomy</h2>
<p>Then additional data source can be fully integrated. Finding the lineage of the organism is as simple as:</p>
<pre class="bash"><code>bio ncov --taxon --lineage</code></pre>
<pre><code>superkingdom, Viruses, 10239
   clade, Riboviria, 2559587
      kingdom, Orthornavirae, 2732396
         phylum, Pisuviricota, 2732408
            class, Pisoniviricetes, 2732506
               order, Nidovirales, 76804
                  suborder, Cornidovirineae, 2499399
                     family, Coronaviridae, 11118
                        subfamily, Orthocoronavirinae, 2501931
                           genus, Betacoronavirus, 694002
                              subgenus, Sarbecovirus, 2509511
                                 species, Severe acute respiratory syndrome-related coronavirus, 694009
                                    no rank, Severe acute respiratory syndrome coronavirus 2, 2697049</code></pre>
</div>
<div id="see-the-bioproject" class="section level2">
<h2>See the bioproject</h2>
<p><code>bio</code> knows about bioprojects and sequencing data, what is stored in the SRA about our data?</p>
<pre class="bash"><code>bio ncov --sra</code></pre>
<pre><code>ncov    BioProject  PRJNA485481
ncov    BioSample   </code></pre>
<p>Alas this particular data is not properly cross-referenced even at NCBI, thus we can’t quite get the SRR run numbers automatically. But for other data, that is deposited correctlywe could just do a:</p>
<pre><code>bio ncov --sra --sample</code></pre>
<p>to see the sample information as stored in the BioProject.</p>
</div>
<div id="bio-is-a-data-model" class="section level2">
<h2><code>bio</code> is a data model</h2>
<p>Beyond the functionality that we show, <code>bio</code> is also an exploration into modeling biological data. The current standards and practices are woefully antiquated and painfully inadequate, the default formats at GenBank or EMBL are dishearteningly inefficient, and depressingly tedious to program with.</p>
<p>In contrast, take a look under the hood, in <code>bio</code> all data are in a simple, efficient, quick to load, compressed in JSON format.</p>
<pre class="bash"><code>bio ncov | head -20</code></pre>
<pre><code>[
    {
        &quot;id&quot;: &quot;ncov&quot;,
        &quot;definition&quot;: &quot;Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome&quot;,
        &quot;dblink&quot;: [
            &quot;BioProject:PRJNA485481&quot;
        ],
        &quot;locus&quot;: &quot;NC_045512&quot;,
        &quot;feature_count&quot;: 57,
        &quot;origin_len&quot;: 29903,
        &quot;molecule_type&quot;: &quot;ss-RNA&quot;,
        &quot;topology&quot;: &quot;linear&quot;,
        &quot;data_file_division&quot;: &quot;VRL&quot;,
        &quot;date&quot;: &quot;18-JUL-2020&quot;,
        &quot;accessions&quot;: [
            &quot;NC_045512&quot;
        ],
        &quot;sequence_version&quot;: 2,
        &quot;keywords&quot;: [
            &quot;RefSeq&quot;</code></pre>
<p>The data layout allows <code>bio</code> to read in the entire human chromosome 1, with its 253 million characters and 328 thousand genomic features, in just three(!) seconds. In another 3 seconds <code>bio</code> can convert that information fasta or gff, it can filter it by type, translate the sequence, extract proteins, slice by coordinate etc:</p>
<pre><code>time bio chr1 --fasta | wc -c
253105766

real    0m6.238s
user    0m4.156s
sys     0m2.172s</code></pre>
<p>For shorter genomes, bacterial or viral the conversion times are under a fraction of a second.</p>
<p>Thanks to the representation it is trivially easy to extend <code>bio</code>. The data is already structured in an efficient layout that needs no additional parsing to load.</p>
</div>
<div id="what-did-bio-do-for-us" class="section level2">
<h2>What did <code>bio</code> do for us?</h2>
<ol style="list-style-type: decimal">
<li>fetched the data from NCBI</li>
<li>created a more efficient local representation the data</li>
<li>stored this representation so that next time you need it is available much faster</li>
<li>generated alignments</li>
</ol>
</div>
<div id="but-wait-there-is-more" class="section level2">
<h2>But wait there is more</h2>
<p>How about translating the reverse of the last 10 nucleotides of every feature labeled as <code>CDS</code>. <code>bio</code> can do that, just keep adding parameters :-) :</p>
<pre><code>bio ncov --fasta --type CDS --start -10 --reverse --translate</code></pre>
<p>ah yes, if that is what you needed, here it is:</p>
<pre><code>&gt;YP_009724389.1 [-9:21291], reverse, translated DNA
NQQ

&gt;YP_009725295.1 [-9:13218], reverse, translated DNA
NVA

&gt;YP_009724390.1 [-9:3822], reverse, translated DNA
NTH
...</code></pre>
<p>And so on. <code>bio</code> has a wealth of utility that makes bioinformatics more accessible.</p>
</div>
<div id="comparisons-to-emboss" class="section level2">
<h2>Comparisons to EMBOSS</h2>
<p>The software with the most similar goals to <code>bio</code> is the <a href="http://emboss.sourceforge.net/">emboss suite</a>, a revolutionary software package developed decades ahead of its time. Unfortunately, perhaps because of being developed so early on, the amazing feats of software engineering within <code>emboss</code> are deployed with a nearly incomprehensible documentation that attempts, in vain, to describe an incredibly obtuse command interface.</p>
<p>We love the concept of <code>emboss</code> but even after many years we don’t understand how to use it. We constantly have to consult the manual for details. Moreover commands that use <code>emboss</code> suites tend to end up as a series of hard to read arcane commands that are surprisingly difficult to comprehend even for experienced scientists.</p>
<p>Criticism aside, imitation is the greatest form of flattery, <code>bio</code> is an homage to <code>emboss</code> with the hope that one day, we can replace some functionality from <code>emboss</code> with code that brings joy rather than frustrations.</p>

</div>
</div>

<script src="assets/asciinema-player.js"></script>

            </section>

          </div>
        </div>
      </div>

<a href="bio-install.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": null,
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "none"
}
});
});
</script>

</body>

</html>
